src/LevelZeroShimImp.hpp

```diff
@@ -0,0 +1,136 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef LEVELZEROSHIMIMP_HPP_INCLUDE
+#define LEVELZEROSHIMIMP_HPP_INCLUDE
+
+#include <string>
+
+#include <level_zero/ze_api.h>
+#include <level_zero/zes_api.h>
+
+#include "LevelZeroShim.hpp"
+
+#include "geopm_time.h"
+
+namespace geopm
+{
+    class LevelZeroShimImp : public LevelZeroShim
+    {
+        public:
+            LevelZeroShimImp();
+            virtual ~LevelZeroShimImp() = default;
+            virtual int num_accelerator(void) const override;
+            virtual int num_accelerator_subdevice(void) const override;
+
+            virtual int frequency_domain_count(unsigned int device_idx, int domain) const override;
+            virtual double frequency_status(unsigned int device_idx, int domain, int domain_idx) const override;
+            virtual double frequency_min(unsigned int device_idx, int domain, int domain_idx) const override;
+            virtual double frequency_max(unsigned int device_idx, int domain, int domain_idx) const override;
+
+            virtual int engine_domain_count(unsigned int device_idx, int domain) const override;
+            virtual uint64_t active_time(unsigned int device_idx, int domain, int domain_idx) const override;
+            virtual uint64_t active_time_timestamp(unsigned int device_idx, int domain, int domain_idx) const override;
+
+            virtual uint64_t energy(unsigned int device_idx) const override;
+            virtual uint64_t energy_timestamp(unsigned int device_idx) const override;
+            virtual int32_t power_limit_tdp(unsigned int device_idx) const override;
+            virtual int32_t power_limit_min(unsigned int device_idx) const override;
+            virtual int32_t power_limit_max(unsigned int device_idx) const override;
+
+            virtual void frequency_control(unsigned int device_idx, int domain, int domain_idx, double setting) const override;
+
+        private:
+            struct frequency_s {
+                double voltage = 0;
+                double request  = 0;
+                double tdp = 0;
+                double efficient = 0;
+                double actual = 0;
+                uint64_t throttle_reasons = 0;
+            };
+            struct power_limit_s {
+                int32_t tdp = 0;
+                int32_t min= 0;
+                int32_t max = 0;
+            };
+
+            virtual void domain_cache(unsigned int device_idx);
```

Comment:
```suggestion
            void domain_cache(unsigned int device_idx);
```
These private funcs don't need to be virtual.